### PR TITLE
#2651 introduce SMB source connector

### PR DESCRIPTION
**Pull request by eechih**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2651)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbFileSystem.scala [hunk 1]
```diff
@@ -21,11 +21,11 @@ import com.island.ohara.common.util.CommonUtils
 
 class TestSmbFileSystem extends FileSystemTestBase {
 
-  private[this] val hostname: Option[String] = sys.env.get(SMB_HOSTNAME_KEY)
-  private[this] val port: Option[String] = sys.env.get(SMB_PORT_KEY)
-  private[this] val user: Option[String] = sys.env.get(SMB_USERNAME_KEY)
-  private[this] val password: Option[String] = sys.env.get(SMB_PASSWORD_KEY)
-  private[this] val shareName: Option[String] = sys.env.get(SMB_SHARE_NAME_KEY)
+  private[this] val hostname: Option[String] = sys.env.get(IT_SMB_HOSTNAME_KEY)
```
Comment (chia7712) [inline ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbFileSystem.scala]: seems this parser is duplicate to TestSmbSource. could we reuse it?

File: ohara-connector/src/main/scala/com/island/ohara/connector/smb/package.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    .documentation("The hostname of SMB server")
+    .valueType(SettingDef.Type.STRING)
+    .build()
+
+  val SMB_PORT_KEY: String = "smb.port"
+  val SMB_PORT_DEFINITION: SettingDef = SettingDef
+    .builder()
+    .key(SMB_PORT_KEY)
+    .documentation("The port of SMB server")
+    .valueType(SettingDef.Type.INT)
```
Comment (chia7712) [inline ohara-connector/src/main/scala/com/island/ohara/connector/smb/package.scala]: Type.INT -> Type.PORT

Comment (eechih) [reply ohara-connector/src/main/scala/com/island/ohara/connector/smb/package.scala]: I will address it. :)

File: ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbFileSystem.scala (hunk 1 above)
Comment (eechih) [reply ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbFileSystem.scala]: 我觀察到 IT 參數好像有一個命名規則 ohara.it.***（難道是個人錯覺~~），所以想保持一致性。不過，這會跟 connector props 的參數不一致，最後才會選擇分開定義。

Comment (chia7712) [reply ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbFileSystem.scala]: 我指的是你建立file system的流程似乎差不多 相關的程式碼可否重複使用？

Comment (eechih) [reply ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbFileSystem.scala]: 喔喔～我誤會了。 I will refactor it.

**Comment by eechih** [comment]
@chia7712 address your comments, please review again, thanks.

File: ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbSource.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.island.ohara.client.filesystem.FileSystem
+import com.island.ohara.connector.CsvSourceTestBase
+import com.island.ohara.connector.smb._
+import com.island.ohara.kafka.connector.csv.CsvConnector._
+import com.island.ohara.kafka.connector.csv.CsvSourceConnector
+
+class TestSmbSource extends CsvSourceTestBase {
+
+  private[this] var itProps: ITSmbProps = _
```
Comment (chia7712) [inline ohara-it/src/test/scala/com/island/ohara/it/connector/smb/TestSmbSource.scala]:
```scala
private[this] var itProps: ITSmbProps = try ITSmbProps(sys.env)
  catch {
    case e: IllegalArgumentException => throw new AssumptionViolatedException(s"skip TestSmbSource test, ${e.getMessage}")
  }
```

**Comment by eechih** [comment]
retry

**Pull request merged**